### PR TITLE
Turn on name mapping for avro

### PR DESCRIPTION
**Pull request by chenjunjiedada**

### CODE REVIEW

**Comment by rdblue** [comment]
Can you create a test case for this with an Avro file written without Iceberg helpers that is read using a name mapping?

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 1]
```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
...
+    NameMapping nameMapping = MappingUtil.create(filteredSchema);
+
+    Table table = catalog.createTable(
+        org.apache.iceberg.catalog.TableIdentifier.of(DB_NAME, "avro_table"),
+        filteredSchema,
+        PartitionSpec.unpartitioned());
+
+    table.updateProperties()
+        .set(DEFAULT_NAME_MAPPING, NameMappingParser.toJson(nameMapping))
+        .set(DEFAULT_FILE_FORMAT, "avro")
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:359]: FYI, this isn't necessary because the files added to a table don't have to match the format.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 2]
```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
...
+        .commit();
+
+    table.newFastAppend().appendFile(avroDataFile).commit();
+
+    List<String> actual = spark.read().format("iceberg")
+        .load(DB_NAME + ".avro_table")
+        .select("name")
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:366]: This should not project because ID is being ignored.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 3]
```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
...
+        .withFileSizeInBytes(avroFile.length())
+        .withPath(avroFile.getAbsolutePath())
+        .withRecordCount(2)
+        .build();
+
+    Schema filteredSchema = new Schema(
+        required(1, "name", Types.StringType.get())
+    );
+
+    NameMapping nameMapping = MappingUtil.create(filteredSchema);
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:350]: I like that `filteredSchema` is used to create the mapping. Could you use a full schema for the table (1: name, 2: id) so that we can verify that the ID is projected as null?

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 4]
```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
...
+        .select("name")
+        .filter("name='Alice'")
+        .collectAsList()
+        .stream()
+        .map(r -> r.getString(0))
+        .collect(Collectors.toList());
+
+    List<GenericRecord> expected = Lists.newArrayList(record2);
+
+    Assert.assertEquals(expected.stream().map(r -> r.get("name")).collect(Collectors.toList()), actual);
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:375]: I think this validation looks strange. There's no need to create a list with just record2, and no need to transform that list. It is easier to read and maintain tests with simple assertions, like `assertEquals("Should project 1 record", 1, actual.size());`. And each field should have its own assertion since you're hard-coding the filter for `Alice` and we know that id will be null.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java [hunk 5]
```diff
@@ -298,4 +309,69 @@ public void testImportWithNameMappingForVectorizedParquetReader() throws Excepti
 
     Assert.assertEquals(expected.stream().map(SimpleRecord::getData).collect(Collectors.toList()), actual);
   }
+
+  @Test
+  public void testAvroReaderWithNameMapping() throws IOException {
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:314]: Why was this added to `SparkTableUtil`? It doesn't use `SparkTableUtil`, so it is not related.

Could you start a test suite, `TestNameMappingProjection`?

Comment (chenjunjiedada) [reply spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:314]: will do.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java (hunk 3 above)
Comment (chenjunjiedada) [reply spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java:350]: We can not use a full schema for the table because it cannot build Avro read schema (with `name` and `id` column) by using pruned schema (with `name` column) in `ProjectionDatumReader#setSchema`.

Also, `setSchema` will ignore the name mapping if the Avro file schema has Ids. So we might need to utilize the `RemoveIds` from the Avro test via moving it to source dir and changing it to the public class. What do you think about this?

Comment (rdblue) [reply spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java]: Can you clarify what you mean by "it cannot build Avro read schema by using pruned schema"? We should be able to use any table schema here, and without a mapping for id, it should be ignored.

We added special handling for this case where the table schema and the file schema use the same name, but the column should not be projected because this ID is missing.

> setSchema will ignore the name mapping if the Avro file schema has Ids

Isn't this why the Avro file schema is created without IDs and no Iceberg classes are used for the write? I don't see why we would have IDs in the Avro file.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestNameMappingProjection.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        .build();
+
+    Schema filteredSchema = new Schema(
+        required(1, "name", Types.StringType.get())
+    );
+    NameMapping nameMapping = MappingUtil.create(filteredSchema);
+
+    Table table = catalog.createTable(
+        org.apache.iceberg.catalog.TableIdentifier.of(DB_NAME, "avro_table"),
+        filteredSchema,
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestNameMappingProjection.java:127]: This is the line that I would change, so that the table also has an ID column.

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestNameMappingProjection.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+
+    table.newFastAppend().appendFile(avroDataFile).commit();
+
+    List<Row> actual = spark.read().format("iceberg")
+        .load(DB_NAME + ".avro_table")
+        .filter("name='Alice'")
+        .collectAsList();
+
+    Assert.assertEquals("Should project 1 record", 1, actual.size());
+    Assert.assertEquals("Should equal to 'Alice'", "Alice", actual.get(0).getString(0));
```
Comment (rdblue) [inline spark2/src/test/java/org/apache/iceberg/spark/source/TestNameMappingProjection.java:142]: And we would need a second assertion here that the `id` value is null.

**Comment by rdblue** [comment]
@chenjunjiedada, I think this is good to commit. It would be nice to update the tests, but not really necessary to test what this is doing. Thanks!

File: spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java (hunk 3 above)
Comment (chenjunjiedada) [reply spark2/src/test/java/org/apache/iceberg/spark/source/TestSparkTableUtil.java]: This is the stack trace that when using a full schema for the table:

>
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0, localhost, executor driver): java.lang.IllegalArgumentException: Missing required field: id 
at org.apache.iceberg.relocated.com.google.common.base.Preconditions.checkArgument(Preconditions.java:217)
at org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:96)
at org.apache.iceberg.avro.BuildAvroProjection.record(BuildAvroProjection.java:41)
at org.apache.iceberg.avro.AvroCustomOrderSchemaVisitor.visit(AvroCustomOrderSchemaVisitor.java:51) 
at org.apache.iceberg.avro.AvroSchemaUtil.buildAvroProjection(AvroSchemaUtil.java:104)
at org.apache.iceberg.avro.ProjectionDatumReader.setSchema(ProjectionDatumReader.java:60)
at org.apache.avro.file.DataFileStream.initialize(DataFileStream.java:132)
at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:106)
at org.apache.avro.file.DataFileReader.<init>(DataFileReader.java:98)
at org.apache.avro.file.DataFileReader.openReader(DataFileReader.java:66)                                                                                                                                          at org.apache.iceberg.avro.AvroIterable.newFileReader(AvroIterable.java:94)
at org.apache.iceberg.avro.AvroIterable.iterator(AvroIterable.java:77)
at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:106)                                                                      at org.apache.iceberg.spark.source.BaseDataReader.next(BaseDataReader.java:73)
at org.apache.spark.sql.execution.datasources.v2.DataSourceRDD$$anon$1.hasNext(DataSourceRDD.scala:49)
at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37) 
at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
...

So that I didn't use the full schema and assert the id column.

>Isn't this why the Avro file schema is created without IDs and no Iceberg classes are used for the write? I don't see why we would have IDs in the Avro file.

Yes, agree to create Avro file without IDs. Here I just want to use `RemoveIds` class from Iceberg.

**Comment by chenjunjiedada** [comment]
@rdblue, Thanks for merging, will think more about the requirement next time.

**Comment by rdblue** [comment]
@chenjunjiedada, the problem was that the ID column was required. It should be optional if it is missing in a file. Could you submit a PR with the test updated?

**Comment by chenjunjiedada** [comment]
@rdblue, Just created https://github.com/apache/iceberg/pull/1135.
